pkg/blockio/blockio.go

```diff
@@ -0,0 +1,456 @@
+	}
+
+	classBlockIO = map[string]cgroups.BlockIOParameters{}
+	// Create cgroup blockio parameters for each blockio class
+	for class := range opt.Classes {
+		cgBlockIO, err := devicesParametersToCgBlockIO(opt.Classes[class], currentIOSchedulers)
+		if err != nil {
+			if force {
+				log.Warnf("ignoring: %v", err)
+			} else {
+				return err
+			}
+		}
+		classBlockIO[class] = cgBlockIO
+	}
+	return nil
+}
+
+// GetClasses returns block I/O class names
+func GetClasses() []string {
+	classNames := make([]string, 0, len(classBlockIO))
+	for name := range classBlockIO {
+		classNames = append(classNames, name)
+	}
+	sort.Strings(classNames)
+	return classNames
+}
+
+// SetCgroupClass sets cgroup blkio controller parameters to match
+// blockio class. group is the cgroup directory of the container
+// without mountpoint and controller (blkio) directories. For example:
+// "/kubepods/burstable/POD_ID/CONTAINER_ID"
+func SetCgroupClass(group string, class string) error {
+	cgBlockIO, ok := classBlockIO[class]
+	if !ok {
+		return fmt.Errorf("no BlockIO parameters for class %#v", class)
+	}
+	err := cgroups.ResetBlkioParameters(group, cgBlockIO)
+	if err != nil {
+		return fmt.Errorf("assigning container in cgroup %q to class %#v failed: %w", group, class, err)
+	}
+	return nil
+}
+
+// getCurrentIOSchedulers returns currently active I/O scheduler used for each block device in the system.
+// Returns schedulers in a map: {"/dev/sda": "bfq"}
+func getCurrentIOSchedulers() (map[string]string, error) {
+	var ios = map[string]string{}
+	schedulerFiles, err := filepath.Glob(sysfsBlockDeviceIOSchedulerPaths)
+	if err != nil {
+		return ios, fmt.Errorf("error in I/O scheduler wildcards %#v: %w", sysfsBlockDeviceIOSchedulerPaths, err)
+	}
+	for _, schedulerFile := range schedulerFiles {
+		devName := strings.SplitN(schedulerFile, "/", 5)[3]
+		schedulerDataB, err := ioutil.ReadFile(schedulerFile)
+		if err != nil {
+			// A block device may be disconnected.
+			log.Errorf("failed to read current I/O scheduler %#v: %v\n", schedulerFile, err)
+			continue
+		}
+		schedulerData := strings.Trim(string(schedulerDataB), "\n")
+		currentScheduler := ""
+		if strings.IndexByte(schedulerData, ' ') == -1 {
+			currentScheduler = schedulerData
+		} else {
+			openB := strings.Index(schedulerData, "[")
+			closeB := strings.Index(schedulerData, "]")
+			if -1 < openB && openB < closeB {
+				currentScheduler = schedulerData[openB+1 : closeB]
+			}
+		}
+		if currentScheduler == "" {
+			log.Errorf("could not parse current scheduler in %#v\n", schedulerFile)
+			continue
+		}
+
+		ios["/dev/"+devName] = currentScheduler
+	}
+	return ios, nil
+}
+
+// deviceParametersToOci converts single blockio class parameters into OCI BlockIO structure.
+func devicesParametersToCgBlockIO(dps []DevicesParameters, currentIOSchedulers map[string]string) (cgroups.BlockIOParameters, error) {
+	var errors *multierror.Error
+	oci := cgroups.NewBlockIOParameters()
```

Comment:
Should this be renamed to something else than `oci`?

Reply:
Definitely, thanks! Fixed.